packages/vite/src/node/server/middlewares/proxy.ts

```diff
@@ -78,40 +87,101 @@ export function proxyMiddleware(
   }
 
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
-  return function viteProxyMiddleware(req, res, next) {
+  return async function viteProxyMiddleware(req, res, next) {
     const url = req.url!
     for (const context in proxies) {
       if (
         (context.startsWith('^') && new RegExp(context).test(url)) ||
         url.startsWith(context)
       ) {
         const [proxy, opts] = proxies[context]
-        const options: HttpProxy.ServerOptions = {}
 
         if (opts.bypass) {
           const bypassResult = opts.bypass(req, res, opts)
           if (typeof bypassResult === 'string') {
             req.url = bypassResult
             debug(`bypass: ${req.url} -> ${bypassResult}`)
             return next()
-          } else if (typeof bypassResult === 'object') {
-            Object.assign(options, bypassResult)
-            debug(`bypass: ${req.url} use modified options: %O`, options)
-            return next()
           } else if (bypassResult === false) {
             debug(`bypass: ${req.url} -> 404`)
             return res.end(404)
           }
         }
 
+        const activeProxyOptions = await prepareProxyRequest(req, opts)
         debug(`${req.url} -> ${opts.target || opts.forward}`)
-        if (opts.rewrite) {
-          req.url = opts.rewrite(req.url!)
-        }
-        proxy.web(req, res, options)
+        proxy.web(req, res, activeProxyOptions)
         return
       }
     }
     next()
   }
 }
+
+async function prepareProxyRequest(
+  req: http.IncomingMessage,
+  opts: ProxyOptions
+) {
+  // req.url = req.originalUrl || req.url
+  const newProxyOptions = Object.assign({}, opts)
+
+  if (opts.router) {
+    const newTarget = await getTarget(req, opts)
+    if (newTarget) {
+      debug('[proxy] Router new target: %s -> "%s"', opts.target, newTarget)
+      newProxyOptions.target = newTarget
+    }
+  }
+  if (opts.rewrite) {
+    req.url = opts.rewrite(req.url!)
```

Comment:
**question:** Why is it safe to use non-null assertion here?
Or could we just add the `req.url` to the if condition and benefit from source flow analysis?

Reply:
It's the same  code with the main branch's code. I just move it to another line.

https://github.com/vitejs/vite/blob/e47a6f2516035643663f9c8260c1f925fc0ccbe0/packages/vite/src/node/server/middlewares/proxy.ts#L82

If req.url are null, the proxy won't run and the process will go into next middleware. Running in dev-server's proxy, I think it's just fine.